File: python/src/aegis/resources/ops.py
"""Ops resource — DevSecOps EvalOps and RedTeam endpoints."""

from __future__ import annotations

from typing import Any, Dict, List

from aegis.models.ops import (
    AttackLibraryEntry,
    BenchmarkResult,
    CiGateRequest,
    CiGateResponse,
    RedTeamStats,
)
from aegis.resources._base import AsyncResource, SyncResource


class SyncOps(SyncResource):

    def ci_gate(self, test_suite: str, **kwargs: Any) -> CiGateResponse:
        req = CiGateRequest(test_suite=test_suite, **kwargs)
        data = self._post("/ops/evalops/ci-gate", json=req.model_dump(exclude_none=True))
        return CiGateResponse.model_validate(data)

    def benchmark(self, benchmark_name: str, **kwargs: Any) -> BenchmarkResult:
        data = self._post(f"/ops/evalops/benchmark/{benchmark_name}", json=kwargs)
        return BenchmarkResult.model_validate(data)

    def get_thresholds(self) -> Dict[str, Any]:
        return self._get("/ops/evalops/thresholds")

    def set_thresholds(self, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/ops/evalops/thresholds", json=kwargs)

    def github_action_check(self, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/ops/evalops/github-action-check", json=kwargs)

    def redteam_regression(self, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/ops/redteam/rr-regression", json=kwargs)

    def redteam_evolution_cycle(self, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/ops/redteam/evolution-cycle", json=kwargs)

    def redteam_stats(self) -> RedTeamStats:
        return RedTeamStats.model_validate(self._get("/ops/redteam/stats"))

    def redteam_continuous_evolution(self, **kwargs: Any) -> Dict[str, Any]:
        return self._post("/ops/redteam/continuous-evolution", json=kwargs)

    def attack_library(self) -> List[AttackLibraryEntry]:
        data = self._get("/ops/redteam/attack-library")
        items = data if isinstance(data, list) else data.get("attacks", [])
        return [AttackLibraryEntry.model_validate(a) for a in items]

    def attack_library_by_category(self, category: str) -> List[AttackLibraryEntry]:
        data = self._get(f"/ops/redteam/attack-library/{category}")
        items = data if isinstance(data, list) else data.get("attacks", [])
        return [AttackLibraryEntry.model_validate(a) for a in items]


class AsyncOps(AsyncResource):

    async def ci_gate(self, test_suite: str, **kwargs: Any) -> CiGateResponse:
        req = CiGateRequest(test_suite=test_suite, **kwargs)
        data = await self._post(
            "/ops/evalops/ci-gate", json=req.model_dump(exclude_none=True)
        )
        return CiGateResponse.model_validate(data)

    async def benchmark(self, benchmark_name: str, **kwargs: Any) -> BenchmarkResult:
        data = await self._post(f"/ops/evalops/benchmark/{benchmark_name}", json=kwargs)
        return BenchmarkResult.model_validate(data)

    async def get_thresholds(self) -> Dict[str, Any]:
        return await self._get("/ops/evalops/thresholds")

    async def set_thresholds(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/ops/evalops/thresholds", json=kwargs)

    async def github_action_check(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/ops/evalops/github-action-check", json=kwargs)

    async def redteam_regression(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/ops/redteam/rr-regression", json=kwargs)

    async def redteam_evolution_cycle(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/ops/redteam/evolution-cycle", json=kwargs)

    async def redteam_stats(self) -> RedTeamStats:
        return RedTeamStats.model_validate(await self._get("/ops/redteam/stats"))

    async def redteam_continuous_evolution(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._post("/ops/redteam/continuous-evolution", json=kwargs)

    async def attack_library(self) -> List[AttackLibraryEntry]:
        data = await self._get("/ops/redteam/attack-library")
        items = data if isinstance(data, list) else data.get("attacks", [])
        return [AttackLibraryEntry.model_validate(a) for a in items]

    async def attack_library_by_category(self, category: str) -> List[AttackLibraryEntry]:
        data = await self._get(f"/ops/redteam/attack-library/{category}")
        items = data if isinstance(data, list) else data.get("attacks", [])
        return [AttackLibraryEntry.model_validate(a) for a in items]

File: go/ops.go
package aegis

import (
	"context"

	"github.com/ai-aegis-yatav/aegis-sdk/go/internal"
	"github.com/ai-aegis-yatav/aegis-sdk/go/models"
)

type OpsService struct {
	t *internal.Transport
}

func (s *OpsService) CiGate(ctx context.Context, req models.CiGateRequest) (*models.CiGateResponse, error) {
	var resp models.CiGateResponse
	if err := s.t.Do(ctx, "POST", "/v1/ops/ci-gate", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OpsService) Benchmark(ctx context.Context, req models.BenchmarkRequest) (*models.BenchmarkResponse, error) {
	var resp models.BenchmarkResponse
	if err := s.t.Do(ctx, "POST", "/v1/ops/benchmark", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OpsService) GetThresholds(ctx context.Context) (*models.ThresholdsConfig, error) {
	var resp models.ThresholdsConfig
	if err := s.t.Do(ctx, "GET", "/v1/ops/thresholds", nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OpsService) SetThresholds(ctx context.Context, req models.ThresholdsConfig) (*models.ThresholdsConfig, error) {
	var resp models.ThresholdsConfig
	if err := s.t.Do(ctx, "PUT", "/v1/ops/thresholds", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OpsService) RedteamStats(ctx context.Context) (*models.RedteamStats, error) {
	var resp models.RedteamStats
	if err := s.t.Do(ctx, "GET", "/v1/ops/redteam/stats", nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

func (s *OpsService) AttackLibrary(ctx context.Context, params models.ListParams) (*models.PaginatedResponse[models.AttackEntry], error) {
	var resp models.PaginatedResponse[models.AttackEntry]
	path := buildPath("/v1/ops/redteam/attacks", params.ToQuery())
	if err := s.t.Do(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}
